### PR TITLE
Add scientific notation support to cudf::strings::to_fixed_point and cudf::strings::is_fixed_point

### LINKED ISSUE #7665: [FEA] cudf::strings::to_fixed_point support scientific notation
**Is your feature request related to a problem? Please describe.**
Currently, `cudf::strings::to_fixed_point` can not parse input string formatted as scientific notation, such as: `1.76e-3`.  And scientific notation is widely used in Spark and other kinds of applications. 

**Describe the solution you'd like**
Reinforce `string_to_decimal_fn`.

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #7665 
This adds support for converting to fixed-point types from strings containing scientific notation like "1.23E2" and "45.6e-12"
This PR also corresponding updates to the `cudf::strings::is_fixed_point` function as well.
The gtests for these APIs were also updated to include valid and invalid strings as appropriate.

Overall, this was a rewrite of the current internal functors for converting and data validation. Also, common code for both were refactored into an internal base class to minimize maintenance.

### CODE REVIEW

File: cpp/src/strings/convert/convert_fixed_point.cu [hunk 1]
```diff
@@ -33,74 +33,152 @@
...
+   *
+   * @tparam check_only Set to true to verify the characters are valid and the
+   *         exponent value in the decimal string does not overflow int32
+   * @param[in,out] iter Start of characters to parse
+   *                     (points to the character after the 'E' or 'e')
+   * @param[in] end End of characters to parse
+   * @return Integer value of the exponent
+   */
+  template <bool check_only = false>
+  __device__ thrust::optional<int> parse_exponent(char const* iter, char const* iter_end) const
```
Comment (hyperbolic2346) [inline cpp/src/strings/convert/convert_fixed_point.cu:102]: Should you call out that this int is 32-bit? I know this is pedantic at best though.

**Review by hyperbolic2346 — APPROVED** [review]
I really like this change. The consolidation of common code is great here. Still seems a good deal of duplication, but certainly a step in the correct direction.

File: cpp/src/strings/convert/convert_fixed_point.cu (hunk 1 above)
Comment (davidwendt) [reply cpp/src/strings/convert/convert_fixed_point.cu:102]: Yes, I think you are right. I should change these to `int32_t` to make it more clear.

File: cpp/src/strings/convert/convert_fixed_point.cu [hunk 2]
```diff
@@ -33,74 +33,152 @@
...
+    int exp_offset     = 0;
+    bool decimal_found = false;
+
+    while (iter < iter_end) {
+      auto const ch = *iter++;
+      if (ch == '.' && !decimal_found) {
+        decimal_found = true;
+        continue;
+      }
+      if (ch < '0' || ch > '9') {
```
Comment (codereport) [inline cpp/src/strings/convert/convert_fixed_point.cu:72]: Is there a device version of `isdigit`?
```suggestion
      if (not isdigit(ch)) {
```

**Review by codereport — APPROVED** [review]

File: cpp/src/strings/convert/convert_fixed_point.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/strings/convert/convert_fixed_point.cu:72]: No, there is no device version of `std::isdigit()`.

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**